tket/src/ArchAwareSynth/SteinerForest.cpp

```diff
@@ -351,88 +349,121 @@ Circuit phase_poly_synthesis(
-  }
+  return PhasePolyBox(circuit_ppb_place);
+}
 
-  if ((cnottype == CNotSynthType::Rec) || (cnottype == CNotSynthType::SWAP)) {
-    node_order = iter_order.get_iterationorder();
+class PhasePolySynthesizer {
+ public:
+  PhasePolySynthesizer(
+      const Architecture &arch, const PhasePolyBox &phasepolybox,
+      unsigned lookahead, CNotSynthType cnottype)
+      : arch(arch),
+        placed_ppb(make_placed_ppb(arch, phasepolybox)),
+        lookahead(lookahead),
+        cnottype(cnottype) {}
+
+  Circuit get_result() {
+    return (cnottype == CNotSynthType::HamPath) ? get_result_using_hampath()
+                                                : get_result_standard();
   }
 
-  unsigned counter = 0;
-  if (cnottype == CNotSynthType::HamPath) {
-    for (Node orig_node : hampath) {
-      UnitID qu_no = UnitID(orig_node);
-      Qubit q = Qubit(counter);
-      Node n = Node(counter);
+ private:
+  Circuit get_result_from_hampath(const std::vector<Node> &hampath) {
+    // maps from qubits/node to int
+    std::map<UnitID, UnitID> forward_contiguous_uids_q;
+    std::map<UnitID, UnitID> backward_contiguous_uids_n;
+    // extra map with node type needed for the creation of the architecture
+    std::map<UnitID, Node> unitid_to_int_nodes;
+    for (unsigned i = 0; i < hampath.size(); i++) {
+      UnitID qu_no = UnitID(hampath[i]);
+      Qubit q = Qubit(i);
+      Node n = Node(i);
       unitid_to_int_nodes.insert({qu_no, n});
       forward_contiguous_uids_q.insert({q, qu_no});
       backward_contiguous_uids_n.insert({qu_no, n});
-
-      ++counter;
     }
-  } else {
-    for (Node orig_node : node_order) {
-      UnitID qu_no =
-          UnitID(orig_node);  // convert node to superclass type of qubit/node
-      Qubit q = Qubit(counter);
-      Node n = Node(counter);
-      forward_contiguous_uids_q.insert({q, qu_no});
-      backward_contiguous_uids_n.insert({qu_no, n});
-      unitid_to_int_nodes.insert({qu_no, n});
-      ++counter;
-    }
-  }
-
-  // define new arcitecture
-  std::vector<Architecture::Connection> new_con;
-  if (cnottype == CNotSynthType::HamPath) {
+    // define new arcitecture
+    std::vector<Architecture::Connection> new_con;
     for (auto pair : arch.get_all_edges_vec()) {
       new_con.push_back(
           {unitid_to_int_nodes[UnitID(pair.first)],
            unitid_to_int_nodes[UnitID(pair.second)]});
     }
-  } else {
-    // include only the tree edges:
+    Architecture con_arch(new_con);
+    return make_result_from_con_arch(
+        con_arch, forward_contiguous_uids_q, backward_contiguous_uids_n);
+  }
+
+  Circuit get_result_using_hampath() {
+    std::vector<Node> hampath = find_hampath(arch);  // using default timeout
+    Circuit c0 = get_result_from_hampath(hampath);
+    // Sometimes the reversed path gives a better circuit. Try both!
+    std::reverse(hampath.begin(), hampath.end());
+    Circuit c1 = get_result_from_hampath(hampath);
+    return (c0.depth() < c1.depth()) ? c0 : c1;
```

Comment:
I'm not sure; I expect they will often give the same answer. I used depth because [this test](https://github.com/CQCL/tket/blob/develop/pytket/tests/architecture_aware_synthesis_test.py#L156) was one that changed from 36 to 28 with the new algorithm; and in the failing long C++ test the two metrics would give the same choice.

There is almost certainly room for more improvement here: for example, returning more than one Hamiltonian path and comparing them. But I wanted to keep it simple for now.

Reply:
Okay, yes I think there is a lot more we could do in the future if we want to.